### DISCUSSION TITLE
关于线程阻塞。

### DISCUSSION THREAD

**Discussion by nanshaobit** [nonebot/discussions#22] [问答] [answered]
在部分任务执行时，较为占有时间，即使使用了`scheduler.add_job`，还是会阻塞线程。

我尝试改用`uvicorn`多`workers`部署时， bot端事件推送正常，程序处理并下发的也是正常的。但是对于某些定时任务，出现了多次下发的情况（多个wokers）。

想了解一下是否有其他更优方案，可以避免或减少阻塞的情况。

**Comment by mnixry** [comment, marked as answer]
如果我没有理解出错的话, 这个阻塞线程问题的应该是在异步函数中采用了同步调用

如果的确是这样, 我们提供了以下工具函数, 通过[调用线程池](https://docs.python.org/zh-cn/3/library/asyncio-eventloop.html#asyncio.loop.run_in_executor)将任意同步步函数调用变为异步, 以避免主线程事件循环被阻塞:
  - 在NoneBot1中, 可以使用[`aiocqhttp.utils.ensure_async`](https://aiocqhttp.nonebot.dev/module/aiocqhttp/utils.html)
  - 在NoneBot2中, 可以使用[`nonebot.utils.run_sync`](https://v2.nonebot.dev/api/utils.html#run-sync-func)

**Reply by nanshaobit** [reply]
是这样的，问题已解决～

多谢。

**Reply by nanshaobit** [reply]
还有一个问题，这个方法无法传递参数吗？

**Reply by mnixry** [reply]
这个函数的推荐用法是作为装饰器使用的
```python
@run_sync
def some_sync_function(a,b,c):
    do_something()

await some_sync_function(1,2,3)
```
如果你不想把它当作装饰器使用, 可以使用下面这种代码
```python
def some_sync_function(a,b,c):
    do_something()

await run_sync(some_sync_function)(1,2,3)
```
~~因为后者看起来很丑所以不推荐~~

**Reply by qllokirin** [reply]
请问如何some_sync_function是个异步函数应该怎么办呢，因为我想在这个函数里用await bot.send_private_msg()回复消息
代码如下
```
import time
#我想在里面使用await bot.send_private_msg() 看编辑器提示所以我加了个async 
@run_sync
async def some_sync_function():
    pass
    #await bot.send_private_msg()
@scheduler.scheduled_job("cron", minute="*/1")
async def every_10_minutes_check():
    await some_sync_function()
```
报错内容
```
RuntimeWarning: coroutine 'some_sync_function' was never awaited
  handle = None  # Needed to break cycles when an exception occurs.
RuntimeWarning: Enable tracemalloc to get the object allocation traceback
```
或者我是不是应该把要发生的消息当做返回值然后在every_10_minutes_check函数里面发生代码应该才是更加规范的？谢谢
（纯纯小白一个，只略懂python基础语法，如有错误还请多多指点）

**Reply by yanyongyu** [reply]
已经是异步函数的话你就不需要用run sync装饰了

**Reply by qllokirin** [reply]
我感觉可能是我对异步的使用方法理解不深刻导致了一些使用错误，我目前的问题是这样的
这是我的代码示例
```
import time
async def run_every_10_minutes_check():
    logger.info("sleep")
    # 这里会执行一些requests请求 我用sleep代替了
    time.sleep(10)
    logger.info("over")

@scheduler.scheduled_job("cron", minute="*/1")
async def every_10_minutes_check():
    await run_every_10_minutes_check()
```
然后情况如下，如果在执行这个定时任务时收到消息，不论是哪个插件的都将无法执行，直到执行完毕（看时间戳）才会接收到消息
我这个`/echo1`是在sleep出现后 over出现前发送的
```
03-04 11:56:00 [INFO] nonebot-plugin-npu | sleep
03-04 11:56:10 [INFO] nonebot-plugin-npu | over
03-04 11:56:10 [INFO] nonebot_plugin_gocqhttp | [xxxx] 收到好友 xxxxxx(xxxx) 的消息: /echo1 (-xxxx)
```
我的理解应该是应该会自动开一个线程去做，不阻塞任何其他的东西，但实际好像不是这样

**Reply by yanyongyu** [reply]
你应该使用`asyncio.sleep`而不是`time.sleep`

**Reply by qllokirin** [reply]
原来如此，非常感谢。我应该把我用的的`requests`函数换成支持异步的如`httpx`或`aiohttp`，谢谢
